# Topics & lectures to be covered

## Week 1
1. Course introduction and general discussion.
2. Molecular biology review

## Week 2
1. Introduction to R / Rstudio.  
   Lecture followed by software setup.
2. Dark matter:  
   Step 1 of practical work based on Amalias mRNA sequences.  
   First give an introduction to how the sequences were obtained and then some
   ideas as to what questions we can ask of them. For details see below.

## Week 3
1. Pairwise alignment (lecture)
2. Implementing pairwise alignment in R (practical)

## Week 4
1. Multiple sequence alignment and trees (lecture)
2. Running multiple sequence alignments in R (practical)

## Week 5
1. Genomes and their structures
2. Looking at genome structure in R (practical)

## Week 6
1. Programming in general
2. Programming exercises in R (practical)

## Week 7
1. Dark matter team work continued
2. Dark matter team work continued

## Week 8
1. Surfing biological databases (lecture)
2. Identifying sequences using online tools (practical)

## Week 9
1. Teamwork presentations 1
2. Teamwork presentations 1

## Week 10
1. Statistics for genomics (lecture)
2. Statistical exercises (practical)

## Week 11
1. Sequence data to p-values: defining differential gene expression (lecture)
2. Using the negative binomial distribution in R (practical)

## Week 12
1. Questions and general discussion (1)
2. Questions and general discussion (2)

## Dark matter practical
This practical is based on the analysis of a set of unknown sequences obtained
by an attempt to directly sequence the mRNA content of 1 gram of benthos. It
will specifically make use of sequences that did not align to known sequences
and to consider how we can consider what these sequences are, including
questions like:

1. Are they real biological sequences or are they somehow experimental artefacts?
2. If real, do they represent:
   1. coding mRNA
   2. other RNA
   3. DNA
   4. Others?
3. Can the sequences tell us anything about the biology of marine mud?

Initially the students will be only given the set of unknown sequences; it is
hoped that they will realise that having other sets of control sequences will
help them. If they do not come up with this, then it may be suggested.

The students will work as groups with each group selecting one members as a
domain specific expert in one of the following fields:
molecular biology, statistics, bioinformatics and graphics. After initial
within group discussion the students will then have discussions in groups of
domain experts; i.e. molecular biology experts will discuss in mol bio
groups. Roughly speaking the individual domain experts will consider the following:

1. **Molecular Biology**  
   The questions that can be used to address the origin of the sequences.
2. **Bioinformatics**  
   How to obtain the information that addresses the questions.
3. **Statistics**  
   How to judge the significance of any particular observations.
4. **Graphics**  
   How to present data.

There is an obvious problem here in that the group one will to some extent set
the question; hence it is important that the set of questions to be addressed
are first raised within the groups and then expanded between the groups.

The practicals will be divided up into different sections with the expectation
that the students will be able to do the following:

### Session 1

Establish a first set of tasks and explore methods of getting these done. To
help with this the students will be provided with some hints.

The exploration of the work should be done directly in R; with questions and
ideas written as comments in the R code where possible, or on paper when it
may be necessary to sketch out ideas of how to visualise data or relationships.

### Session 2

Obtain, visualise and test some numbers relating to the task at hand. Consider
how informative the observations are.

### Session 3

Summarise the observations and prepare a short presentation.

